scipy/ndimage/tests/test_filters.py

```diff
@@ -385,326 +428,450 @@ def test_correlate17(self):
-        kernel = np.array([1, 1 + 1j], dtype)
-        array = np.array([1, 2j, 3, 1 + 4j, 5, 6j], dtype)
-        self._validate_complex(array, kernel, dtype_output, mode='constant',
+                                                       dtype_output, xp):
+        if not (is_numpy(xp) or is_cupy(xp)):
+            pytest.xfail("output=dtype is numpy-specific")
+
+        dtype = getattr(xp, dtype)
+        dtype_output = getattr(xp, dtype_output)
+
+        if is_cupy(xp):
+            pytest.xfail("https://github.com/cupy/cupy/issues/8405")
+
+        kernel = xp.asarray([1, 1 + 1j], dtype=dtype)
+        array = xp.asarray([1, 2j, 3, 1 + 4j, 5, 6j], dtype=dtype)
+        self._validate_complex(xp, array, kernel, dtype_output, mode='constant',
                                cval=5.0 + 2.0j)
 
-    def test_gauss01(self):
-        input = np.array([[1, 2, 3],
-                          [2, 4, 6]], np.float32)
+    def test_gauss01(self, xp):
+        input = xp.asarray([[1, 2, 3],
+                            [2, 4, 6]], dtype=xp.float32)
         output = ndimage.gaussian_filter(input, 0)
         assert_array_almost_equal(output, input)
 
-    def test_gauss02(self):
-        input = np.array([[1, 2, 3],
-                          [2, 4, 6]], np.float32)
+    def test_gauss02(self, xp):
+        input = xp.asarray([[1, 2, 3],
+                            [2, 4, 6]], dtype=xp.float32)
         output = ndimage.gaussian_filter(input, 1.0)
-        assert_equal(input.dtype, output.dtype)
-        assert_equal(input.shape, output.shape)
+        assert input.dtype == output.dtype
+        assert input.shape == output.shape
+
+    def test_gauss03(self, xp):
+        if is_cupy(xp):
+            pytest.xfail("https://github.com/cupy/cupy/issues/8403")
 
-    def test_gauss03(self):
         # single precision data
-        input = np.arange(100 * 100).astype(np.float32)
-        input.shape = (100, 100)
+        input = xp.arange(100 * 100, dtype=xp.float32)
+        input = xp.reshape(input, (100, 100))
         output = ndimage.gaussian_filter(input, [1.0, 1.0])
 
-        assert_equal(input.dtype, output.dtype)
-        assert_equal(input.shape, output.shape)
+        assert input.dtype == output.dtype
+        assert input.shape == output.shape
 
         # input.sum() is 49995000.0.  With single precision floats, we can't
         # expect more than 8 digits of accuracy, so use decimal=0 in this test.
-        assert_almost_equal(output.sum(dtype='d'), input.sum(dtype='d'),
-                            decimal=0)
-        assert_(sumsq(input, output) > 1.0)
-
-    def test_gauss04(self):
-        input = np.arange(100 * 100).astype(np.float32)
-        input.shape = (100, 100)
-        otype = np.float64
+        o_sum = xp.sum(output, dtype=xp.float64)
+        i_sum = xp.sum(input, dtype=xp.float64)
+        assert_almost_equal(o_sum, i_sum, decimal=0)
+        assert sumsq(np.asarray(input), np.asarray(output)) > 1.0
```

Comment:
Ah, indeed! However the test is xfailed on CuPy and has no chance to run under other GPU libs.

Reply:
Sure, but I'm not convinced it is a good idea to write a test which will fail as soon as the xfail reason is resolved (maybe it *won't* get resolved any time soon, but that is besides the point IMO). `sumsq` can easily be written to be array-agnostic:

```python
def sumsq(a, b, xp=None):
    xp = array_namespace(a, b) if xp is None else xp
    return xp.sqrt(xp.sum((a - b)**2))
```

Then call it with `xp=xp` to avoid extra `array_namespace` overhead.

We shouldn't need to be concerned by any slow down from `math.sqrt` to `xp.sqrt` as this is just testing code.